In [ ]:
# Question 2: NumPy Numerical Analysis

## GDP Growth vs Employment by Economic Sector

In [ ]:
## 1. Introduction

#This section performs numerical analysis using NumPy on the cleaned GDP Growth and Employment by Economic Sector dataset. The analysis focuses on applying NumPy arrays and numerical functions to calculate descriptive statistics, perform array operations, reshape data, and identify numerical patterns within the dataset.

#The cleaned dataset used in this analysis was prepared in Question 1 and is stored in the project's `data/clean` folder.

In [3]:
%pip install pandas numpy

   ---------------------------------------- 0.0/10.0 MB ? eta -:--:--
   ---------------------------------------- 0.0/10.0 MB ? eta -:--:--
   ---------------------------------------- 0.0/10.0 MB ? eta -:--:--
   - -------------------------------------- 0.3/10.0 MB ? eta -:--:--
   --- ------------------------------------ 0.8/10.0 MB 2.2 MB/s eta 0:00:05
   ------ --------------------------------- 1.6/10.0 MB 3.2 MB/s eta 0:00:03
   ------------- -------------------------- 3.4/10.0 MB 4.6 MB/s eta 0:00:02
   --------------------- ------------------ 5.2/10.0 MB 5.6 MB/s eta 0:00:01
   ------------------------------ --------- 7.6/10.0 MB 6.7 MB/s eta 0:00:01
   -------------------------------------- - 9.7/10.0 MB 7.3 MB/s eta 0:00:01
   ---------------------------------------- 10.0/10.0 MB 7.0 MB/s  0:00:01
   ---------------------------------------- 0.0/12.6 MB ? eta -:--:--
   ------- -------------------------------- 2.4/12.6 MB 11.7 MB/s eta 0:00:01
   -------------- -----------------

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

[notice] A new release of pip is available: 25.3 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import pandas as pd
import numpy as np

In [3]:
# Load the cleaned dataset
df = pd.read_csv("../../data/clean/GDP_Employment_Cleaned.csv")

df.head()

,REF_AREA_LABEL,TIME_PERIOD,GDP_Growth,Employment_Thousands,Sector
0,Zambia,2015,2.920375,4532.103027,Sector: Total
1,Zambia,2015,2.920375,3089.903564,Sector: Agriculture
2,Zambia,2015,2.920375,99.403816,Sector: Mining
3,Zambia,2015,2.920375,180.855972,Sector: Manufacturing
4,Zambia,2015,2.920375,10.042576,Sector: Utilities


In [6]:
# Display the dataset dimensions
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

# Display column names
print("Columns:")
print(df.columns.tolist())

# Display data types
df.dtypes

Number of rows: 37700
Number of columns: 5
Columns:
['REF_AREA_LABEL', 'TIME_PERIOD', 'GDP_Growth', 'Employment_Thousands', 'Sector']


REF_AREA_LABEL              str
TIME_PERIOD               int64
GDP_Growth              float64
Employment_Thousands    float64
Sector                      str
dtype: object

In [7]:
# Check for missing values in the cleaned dataset
print("Missing values:")
print(df.isnull().sum())

# Check for duplicate rows
print("\nDuplicate rows:", df.duplicated().sum())

Missing values:
REF_AREA_LABEL          0
TIME_PERIOD             0
GDP_Growth              0
Employment_Thousands    0
Sector                  0
dtype: int64

Duplicate rows: 0


## 2. Numerical Data Preparation

The cleaned dataset contains 37,700 observations and five variables. For the NumPy numerical analysis, the numerical variables selected are `TIME_PERIOD`, `GDP_Growth`, and `Employment_Thousands`. The categorical variables `REF_AREA_LABEL` and `Sector` are used to identify countries and economic sectors but are not directly used for numerical calculations.

## 3. NumPy Array Conversion and Descriptive Statistics

In [8]:
# Convert numerical columns to NumPy arrays

years = df["TIME_PERIOD"].to_numpy()
gdp_growth = df["GDP_Growth"].to_numpy()
employment = df["Employment_Thousands"].to_numpy()

print("Years array shape:", years.shape)
print("GDP Growth array shape:", gdp_growth.shape)
print("Employment array shape:", employment.shape)

Years array shape: (37700,)
GDP Growth array shape: (37700,)
Employment array shape: (37700,)


In [9]:
# Display the first 10 values from each NumPy array

print("Years:", years[:10])
print("GDP Growth:", gdp_growth[:10])
print("Employment:", employment[:10])

Years: [2015 2015 2015 2015 2015 2015 2015 2015 2015 2015]
GDP Growth: [2.92037511 2.92037511 2.92037511 2.92037511 2.92037511 2.92037511
 2.92037511 2.92037511 2.92037511 2.92037511]
Employment: [4532.10302734 3089.90356445   99.40381622  180.85597229   10.04257584
   35.70127106  795.23620605   70.21607208   47.9518013   202.79154968]


In [10]:
# Descriptive statistics for GDP Growth

print("GDP Growth Mean:", np.mean(gdp_growth))
print("GDP Growth Median:", np.median(gdp_growth))
print("GDP Growth Minimum:", np.min(gdp_growth))
print("GDP Growth Maximum:", np.max(gdp_growth))
print("GDP Growth Standard Deviation:", np.std(gdp_growth))
print("GDP Growth Variance:", np.var(gdp_growth))

GDP Growth Mean: 4.04977657846982
GDP Growth Median: 3.9534899458060124
GDP Growth Minimum: -50.24806710352907
GDP Growth Maximum: 35.224078310176
GDP Growth Standard Deviation: 4.433296814305468
GDP Growth Variance: 19.65412064373101


In [11]:
# Descriptive statistics for Employment

print("Employment Mean:", np.mean(employment))
print("Employment Median:", np.median(employment))
print("Employment Minimum:", np.min(employment))
print("Employment Maximum:", np.max(employment))
print("Employment Standard Deviation:", np.std(employment))
print("Employment Variance:", np.var(employment))

Employment Mean: 5222.356916579864
Employment Median: 339.66716733067733
Employment Minimum: 0.065
Employment Maximum: 775772.0625
Employment Standard Deviation: 30992.609509177746
Employment Variance: 960541844.1883748


### Interpretation

The NumPy descriptive statistics provide an overview of the distribution of GDP growth and employment values in the cleaned dataset. The mean and median provide measures of central tendency, while the minimum and maximum values show the range of the observations. Standard deviation and variance indicate the amount of variation in the numerical data.

These statistics provide a numerical foundation for understanding the relationship between economic growth and employment across countries, years, and economic sectors.

## 3. NumPy Array Conversion and Descriptive Statistics

The numerical variables were converted from Pandas Series into NumPy arrays to perform numerical analysis. NumPy was used to calculate measures of central tendency and dispersion, including the mean, median, minimum, maximum, standard deviation, and variance.

The GDP Growth results show a mean of approximately 4.05% and a median of approximately 3.95%. The standard deviation of approximately 4.43 indicates variation in GDP growth across the observations.

Employment shows a much larger difference between its mean and median. The mean employment value is approximately 5,222.36 thousand, while the median is approximately 339.67 thousand. This large difference, together with the maximum value of approximately 775,772.06 thousand, indicates substantial variation in employment values across the observations.

In [12]:
# Calculate additional NumPy statistics

gdp_range = np.ptp(gdp_growth)
employment_range = np.ptp(employment)

gdp_percentiles = np.percentile(gdp_growth, [25, 50, 75])
employment_percentiles = np.percentile(employment, [25, 50, 75])

print("GDP Growth Range:", gdp_range)
print("GDP Growth 25th, 50th and 75th Percentiles:", gdp_percentiles)

print("\nEmployment Range:", employment_range)
print("Employment 25th, 50th and 75th Percentiles:", employment_percentiles)

GDP Growth Range: 85.47214541370506
GDP Growth 25th, 50th and 75th Percentiles: [1.93913775 3.95348995 6.11957879]

Employment Range: 775771.9975
Employment 25th, 50th and 75th Percentiles: [  60.16853619  339.66716733 1985.42605586]
